enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,275 @@
+
+## Motivation
+
+To automatically detect local devices and provision local volumes out of them.
+
+### Goals
+
+- Automatic detection of available disks from nodes which can be used as PVs for OpenShift-cluster.
+- Respond to the attach/detach events of the disks/devices.
+- Have options for filtering particular kind of disks based on properties such as name, size, manufacturer, etc.
+
+### Non-Goals
+
+- Inherently protect against confilict with provisioners that own local devices on the same nodes automatic detection is configured to run.
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce a new Custom Resource, `AutoDetectVolumes`, that enables automated discovery and provisioning of local volume based PVs on a set of nodes addressed by one or more labelSelectors.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```go
+
+type AutoDetectVolumeList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+    Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+    Spec              AutoDetectVolumeSpec   `json:"spec"`
+    Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+    // Nodes on which the automatic detection policies must run.
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    // StorageClassName to use for set of matched devices
+    StorageClassName string `json:"storageClassName"`
+    // MinDeviceCount is the minumum number of devices that needs to be detected per node.
+    // If the match devices are less than the minCount specified then no PVs will be created.
+    MinDeviceCount int `json:"minDeviceCount"`
+    // +optional
+    // Maximum number of Devices that needs to be detected per node.
+    MaxDeviceCount int `json:"maxDeviceCount"`
+    // VolumeMode determines whether the PV created is Block or Filesystem
+    // + optional
+    VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+    // FSType type to create when volumeMode is Filesystem
+    // +optional
+    FSType string `json:"fsType,omitempty"`
```

Comment:
I thought we also discussed to have a boolean "dry-run" value which just finds the devices and not really creates them. We are also going to need to design a type that could hold the devices found on the node. We can use configmap or a new type...

Reply:
I disagree with the dry-run value approach. CRs already represent desired state. Using them for dry-run state just doesn't seem right. Discovery should store the available device list somewhere that the UI can load it, present the devices to the user, then create CRs when the user is ready to commit the devices.

Reply:
Okay, but then that still leaves us with an unsolved problem. We want the ability for the UI to trigger a filtering of the set of all discovered devices without 1. Commiting to whatever the result it or 2. Forcing the UI to reimplement the filtering logic of the backend. What's your counter proposal here?

Reply:
The UI is given all the metadata it needs to present the available devices to the user and the UI has to display a reasonable UI around those properties. The filtering itself seems like it would already need to be baked into the UI, such as knowing which nodes hold which devices and how much storage is on each node. 

Where is the filtering complicated enough that we wouldn't have reasonable confidence that the desired state generated in the CR wouldn't end up with the expected PVs?

Reply:
The problem for me lies in the duplication of effort that needs to be maintained in sync by two separate teams that may not be in regular contact with each other. It's not that it's difficult, it's that it's inefficient and more of a long-term maintenance headache than the proposed solution.

Reply:
The maintenance points between the teams would be:
- Discovery info (generated by the LSO discovery and consumed by the UI)
- CRs (generated by the UI and consumed by the LSO)

Where is the duplication of effort? The UI has to implement filtering in the UI widgets and LSO implements filtering to generate the PVs. It seems like very different implementation.

Reply:
There are two sets of information I'm considering:

1. The set of all local devices detected
2. The set of all local devices matching the DeviceDiscoveryPolicy filter

The LSO would need to present set 1 for the UI to present it to the user. Then the user enters filtering information to the UI, which would be in the form of fields that match a DeviceDiscoveryPolicy CR. This would then cause the UI to update and display set 2. 

Here's where I feel like I'm missing something: If the UI does this filtering to generate set 2 on its own it is essentially duplicating the discovery the operator would do based on the final DeviceDiscoveryPolicy CR. This feels like we'd be opening ourselves up to a situation where the operator could change something about how it evaluates the CR without changing the API. Now this is something that has to be duplicates back to the UI so it displays set 2 correctly.

Does that make sense as a concern, at least?

Reply:
The input to the UI is set 1. The output of the UI is the auto discovery CR. Whatever the UI decides to show before it outputs the CR is up to the UI team and is only based on set 1. IMO there is no set 2 unless perhaps the UI finds it useful for some intermediate state before generating the CR. But that's just internal UI implementation. 

If the operator changes some semantics around the CR that impacts the UI, sounds like it's a backward compatibility bug. We would need to test that changes to the CR in the future are compatible with the UI. Beyond this, I don't see a concern.

Reply:
How or what component generates this device list which can be fed to the UI?

Reply:
@jarrpa and I had a discussion on this... Most importantly, the discussion shouldn't block this design PR. Let's keep pushing to get this merged soon. Jose will start a new thread with this topic.

Reply:
>  Most importantly, the discussion shouldn't block this design PR

This discussion is the root of this PR. If the filtering is done in UI, why do we need complicated rules in LSO? UI, knowing which devices user selected / filtered, can create LocalVolume CRs directly, without any `DeviceDiscoveryPolicy`.

Note that output of LSO are Kubernetes PVs to be consumed by OCS (or anything else), not hardware inventory of all disks on a node.

From what I read above, some UI wants to  present hardware inventory and once user selects / filters the devices then UI does *something* so PVs are created. We are discussing the *something* here in this PR, however, I haven't read anything about how the HW inventory gets to UI in the first place. Will there be another daemon running on all nodes collecting information about block devices, just like LSO, only presenting them to Kubernetes in another way than LSO does???

Reply:
@jsafrane The reason we discussed having a separate thread for this topic is that this design doc is independent from the UI. The UI will need to build on top of this new design, but fundamentally this AutoDetectVolume CR doesn't need any UI around it in order to configure the devices. Though if something in this design is blocking the UI design, then I would certainly want to resolve it here.

> From what I read above, some UI wants to present hardware inventory and once user selects / filters the devices then UI does something so PVs are created. We are discussing the something here in this PR

That *something* for the UI to create PVs would be to create the AutoDetectVolume CR(s). This is the signal that the UI is ready to commit to the device selection properties that the user decided on. That *something* could also be for the UI to create individual CRs for each of the devices, and there would be no need for the AutoDetectVolume CR. IMHO, the AutoDetectVolume is a better approach for the UI, but if we settle on the other approach during UI design, then so be it. 

> however, I haven't read anything about how the HW inventory gets to UI in the first place. Will there be another daemon running on all nodes collecting information about block devices, just like LSO, only presenting them to Kubernetes in another way than LSO does???

The thought is that the LSO daemons that are already running on all the nodes would be able to discover the available devices and store their properties *somewhere* for the UI to discover. That *somewhere* could be a ConfigMap for each node, stored as JSON. Or maybe a new CRD type should be created. The CRD seems like overkill since there is nothing for an operator to reconcile. LSO just needs to communicate metadata to the UI.

Reply:
> The thought is that the LSO daemons that are already running on all the nodes would be able to discover the available devices and store their properties somewhere for the UI to discover. 

This was not part of this enhancement. Can we get **all** requirements on LSO into this enhancement, with all API objects and a complete work flow from a "blank" OCP cluster to a cluster with OCS installed (or at least ready to be installed)? Otherwise we may end up with half designed solution that won't make anyone happy.

To me it looks like this PR is trying to reinvent [OpenEBS disk manager](https://github.com/openebs/node-disk-manager/blob/master/docs/design.md)

Reply:
If I were to summarize all the meetings and design discussions...
1. OpenShift admins need a UI to select which devices will be exposed as local PVs
   - The UI needs to display available storage for devices that don't yet have PVs. The UI  (pending UI design) would not show individual devices, but would show some representation of available storage and allow the user to filter devices out of the selection based on ssd/hdd, size, or other properties.
   - The UI would drive LSO to create the local PVs 
   - Different types of devices should be available through different storage classes (ssds vs nvme vs hdd)
1. OCS UI needs to detect what local PVs are available and allow the user to configure OCS to use some subset of the available storage.

We don't see #1 as being under the OCS scope since the local PVs could also be used outside of OCS.

@gnufied Since you've been in most of the discussions, does this match your expectation as well?

@jsafrane There are at least several other solutions in the community for local PV configuration and LSO overlaps with them. Our discussions have been to take a dependency on LSO. Are you suggesting LSO isn't the right place for this? Or simply pointing out a similar design?